core/domain/topic_services.py

```diff
@@ -1686,3 +1686,62 @@ def get_topic_id_to_topic_name_dict(topic_ids: List[str]) -> Dict[str, str]:
         )
         raise Exception(error_msg)
     return topic_id_to_topic_name
+
+
+def update_chapters_counts_in_topic_summaries(
+    topic_summary_dicts: List[topic_domain.FrontendTopicSummaryDict]
+) -> List[topic_domain.FrontendTopicSummaryDict]:
+    """Returns topic summary dicts after updating chapter counts in them.
+
+    Args:
+        topic_summary_dicts: List[FrontendTopicSummaryDict]. A list of
+            topic summary dicts to be updated with chapter counts.
+
+    Returns:
+        List[FrontendTopicSummaryDict]. List of updated topic summary dicts.
+    """
+    topic_ids = [summary['id'] for summary in topic_summary_dicts]
+    topics = topic_fetchers.get_topics_by_ids(topic_ids)
+    all_story_ids: List[str] = []
+    for topic in topics:
+        if topic:
+            story_ids = [story_reference.story_id for
+                story_reference in topic.canonical_story_references]
+            all_story_ids = all_story_ids + story_ids
+    total_stories = story_fetchers.get_stories_by_ids(all_story_ids)
+    for topic in topics:
+        if topic is None:
+            continue
+        topic_summary_dict = next(
+            topic_summary for topic_summary in topic_summary_dicts
+            if topic_summary['id'] == topic.id)
+        upcoming_chapters_count = 0
+        overdue_chapters_count = 0
+        total_chapters_counts = []
+        published_chapters_counts = []
+        stories = [story for story in total_stories if story and
```

Comment:
Don't you already have the story_ids for each topic (you found those in the first for loop)? If so then why not just use those directly here?

Also, when fetching the stories, construct a dict mapping story IDs to story objects so that you can retrieve them easily.

Reply:
Done